integration_tests/README.md

```diff
@@ -303,3 +303,94 @@ The marks you care about are all in marks.py
+   * Numeric Types 
+     * `ByteType` 
+     * `ShortType` 
+     * `IntegerType`
+     * `LongType`
+     * `FloatType`
+     * `DoubleType`
+     * `DecimalType`
+   * Strings 
+     * `StringType` 
+     * `VarcharType`
+   * Binary (`BinaryType`)  
+   * Booleans (`BooleanType`)
+   * Chrono Types 
+     * `TimestampType` 
+     * `DateType`
+     * `Interval`
+   * Complex Types 
+     * `ArrayType`
+     * `StructType`
+     * `MapType`
+
+`data_gen.py` provides `DataGen` classes that help generate test data in integration tests.
+
+### 2. Nested Data types
+Complex data types (`ArrayType`, `StructType`, `MapType`) warrant extensive testing for various combinations of nesting.
+E.g.
+   * `Array<primitive_type>`
+   * `Array<Array<primitive_type>>`
+   * `Array<Struct<primitive_type>>`
+   * `Struct<Array<primitive_type>>`
+   * `Array<Struct<Array<primitive_type>>>`
+   * `Struct<Array<Struct<primitive_type>>>`
+
+The `ArrayGen` and `StructGen` classes in `data_gen.py` can be configured to support arbitrary nesting.
+
+### 3. Scalar values
+Operators and expressions that support scalar operands need to be tested with scalar inputs, of all 
+supported types from 1 and 2, above. 
+For instance, `SUM()` supports numeric columns (e.g. `SUM(a + b)`), or scalars (e.g. `SUM(20)`).
+Similarly, `COUNT()` supports the following:
+   * Columns: E.g. `COUNT(a)` to count non-null rows for column `a`
+   * Scalars: E.g. `COUNT(1)` to count all rows (including nulls)
+   * `*`: E.g. `COUNT(*)`, functionally equivalent to `COUNT(1)`
+It is advised that tests be added for all applicable scalar types, for an operator.
+
+### 4. Null rows
+Ensure that the test data accommodates null values for input columns.
+Null values in input columns are a frequent source of bugs in `rapids-plugin-4-spark`, because of mismatches in null-handling
+and semantics, between RAPIDS `libcudf` (on which `rapids-plugin-4-spark` relies heavily), and Apache Spark. 
+
+Apart from null rows in columns of primitive types, the following conditions must be covered for nested types:
+   * Null rows at the "top" level for `Array`/`Struct` columns.   E.g. `[ [1,2], [3], ∅, [4,5,6] ]`.
+   * Non-null rows containing null elements in the child column. E.g. `[ [1,2], [3,∅], ∅, [4,∅,6] ]`.
+
+### 5. Empty rows in `Array` columns
+Operations on `ArrayType` columns must be tested with input columns containing non-null *empty* rows.
+E.g.
+```
+[
+    [0,1,2,3],
+    [], <------- Empty, non-null row.
+    [4,5,6,7],
+    ...
+]
+```
+
+### 6. Degenerate cases with "empty" inputs
+Ensure that operations are tested with "empty" input columns (i.e. containing zero rows.)
+
+E.g. `COUNT()` on an empty input column yields `0`. `SUM()` yields `0` for the appropriate numeric type.
+
+### 7. Special floating point values
+Apart from `null` values, `FloatType` and `DoubleType` input columns must also include the following special values:
```

Comment:
Should mention that the floating point data generators already cover this by default.

Reply:
Agreed. Mentioned `FloatGen` and `DoubleGen`.